Add PartitionInfo protocol

nicholasbishop opened the pull request:
This is described in section 13.18 of the UEFI specification. It
provides information about MBR or GPT partitions.

josephlr commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+#[repr(packed)]
+#[derive(Clone, Copy, Debug)]
+pub struct GptPartitionEntry {
+    partition_type_guid: Guid,
```
Should we define constants somewhere corresponding to `EFI System Partition` (`C12A7328-F81F-11D2-BA4B-00A0C93EC93B`) and `Partition containing a legacy MBR` (`024DEE41-33E7-11D3-9D69-0008C781F39F`)

nicholasbishop replied:
done, added a new `GptPartitionType` enum for this

josephlr commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+/// GPT partition entry.
```
Nit: "GPT/EFI Partition Entry"

nicholasbishop replied:
done

josephlr commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+/// Legacy MBR partition record.
```
Nit: "Legacy MBR Partition Record"

nicholasbishop replied:
done

josephlr commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+
+impl GptPartitionEntry {
+    /// GUID that defines the type of this Partition. A value of zero
+    /// indicates that this partition entry is used.
```
partition entry is unused.

nicholasbishop replied:
done

josephlr commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+#[repr(C)]
+#[derive(Clone, Copy)]
+union PartitionInfoRecord {
+    /// MBR data.
+    mbr: MbrPartitionRecord,
+
+    /// GPT data.
+    gpt: GptPartitionEntry,
```
no need for doc comments this type is private

nicholasbishop replied:
done

josephlr commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+    ///
+    /// UEFI specification: section 13.18.
+    pub enum PartitionInfoRevision: u32 => {
+        /// Revision of the partition info protocol.
```
Nit: "Revision of `EFI_PARTITION_INFO_PROTOCOL_REVISION`"

nicholasbishop replied:
done

josephlr reviewed (commented):
Looks reasonable to me. However, note that this crate generally doesn't refer to specific sections of the UEFI spec (as they can change from revision to revision), so we should probably drop explicit table/section numbers.

I confirmed that the structs seemed to have the right fields.

nicholasbishop commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+    ending_chs: [u8; 3],
+    starting_lba: u32,
+    size_in_lba: u32,
+}
```
done

nicholasbishop commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+#[derive(Clone, Copy, Protocol)]
+pub struct PartitionInfo {
+    revision: PartitionInfoRevision,
+    partition_type: PartitionType,
```
done

nicholasbishop commented on src/proto/media/partition.rs:
```diff
@@ -0,0 +1,202 @@
+        } else {
+            None
+        }
+    }
```
done, added the revision check as well

nicholasbishop commented on uefi-test-runner/src/proto/media/mod.rs:
```diff
@@ -32,4 +33,23 @@ pub fn test(bt: &BootServices) {
+        let pi = bt
+            .handle_protocol::<PartitionInfo>(handle)
+            .expect_success("Failed to get partition info");
+        let pi = unsafe { &mut *pi.get() };
```
done

nicholasbishop commented:
> Looks reasonable to me. However, note that this crate generally doesn't refer to specific sections of the UEFI spec (as they can change from revision to revision), so we should probably drop explicit table/section numbers.
> 
> I confirmed that the structs seemed to have the right fields.

Thanks for detailed review. I dropped the explicit spec references and I think I've addressed all your comments.

GabrielMajeri commented:
Thanks for the PR, and thanks @josephlr for the excellent review! I've given a quick over the code and it looks good.
